### PR TITLE
Adjust semantic model for method group conversion

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/36377

In the scenario, we have a user-defined conversion for cast, stacked with a method group conversion.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 1]
```diff
@@ -4294,6 +4294,12 @@ private OneOrMany<Symbol> GetMethodGroupSemanticSymbols(
                         // we want to get the symbol that overload resolution chose for M, not the whole method group M.
                         var conversion = (BoundConversion)boundNodeForSyntacticParent;
 
+                        if (conversion.ConversionKind is not ConversionKind.MethodGroup && conversion.Operand is BoundConversion nestedConversion)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:4297]: Nit:
```suggestion
                        if (conversion is { ConversionKind: not ConversionKind.MethodGroup, Operand: BoundConversion nestedConversion })
```

**Review by 333fred — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@jcouv Was this a regression?

File: src/Compilers/CSharp/Test/Symbol/Symbols/ConversionTests.cs [hunk 2]
```diff
@@ -425,6 +426,112 @@ static void Main()
             Assert.True(conversion.IsNumeric);
         }
 
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/36377")]
+        public void GetSymbolInfo_ExplicitCastOnMethodGroup()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/ConversionTests.cs:430]: Consider adding a similar test for VB

File: src/Compilers/CSharp/Test/Symbol/Symbols/ConversionTests.cs [hunk 3]
```diff
@@ -425,6 +426,112 @@ static void Main()
...
+public sealed class C
+{
+    public static void M()
+    {
+        C x = (C)C.Test;
+    }
+
+    public static int Test() => 1;
+
+    public static explicit operator C(System.Func<int> intDelegate)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/ConversionTests.cs:442]: >explicit [](http://example.com/codeflow?start=18&length=8)

Consider testing with implicit operator, including a scenario when the cast operator, `(C)`,  is not present in code.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:4297]: >nestedConversion [](http://example.com/codeflow?start=129&length=16)

It feels like we should also check how this node related to the conversion syntactically

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 4]
```diff
@@ -4294,6 +4294,12 @@ private OneOrMany<Symbol> GetMethodGroupSemanticSymbols(
...
+                        {
+                            Debug.Assert(conversion.ConversionKind is ConversionKind.NoConversion || conversion.ExplicitCastInCode);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:4299]: >Debug.Assert(conversion.ConversionKind is ConversionKind.NoConversion || conversion.ExplicitCastInCode); [](http://example.com/codeflow?start=28&length=104)

It is not obvious why is it reasonable to make this assumption. And more importantly why do we need to make it. If we are going after a specific bound tree shape, I think we should be checking for that instead. For example, we might want to confirm that the nested conversion is the immediate conversion for the ```BoundMethodGroup boundNode```.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 5]
```diff
@@ -4294,6 +4294,12 @@ private OneOrMany<Symbol> GetMethodGroupSemanticSymbols(
...
+                            conversion = nestedConversion;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:4300]: >conversion = nestedConversion; [](http://example.com/codeflow?start=28&length=30)

Consider adding a comment about why/when we need to do this transformation.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

File: src/Compilers/CSharp/Test/Symbol/Symbols/ConversionTests.cs [hunk 6]
```diff
@@ -425,6 +426,207 @@ static void Main()
...
+            var comp = CreateCompilation(src);
+            comp.VerifyEmitDiagnostics(
+                // (5,17): error CS0428: Cannot convert method group 'Test' to non-delegate type 'C'. Did you intend to invoke the method?
+                //         C x = C.Test;
+                Diagnostic(ErrorCode.ERR_MethGrpToNonDel, "Test").WithArguments("Test", "C").WithLocation(5, 17));
+
+            var tree = comp.SyntaxTrees.Single();
+            var model = comp.GetSemanticModel(tree);
+            var memberAccess = GetSyntax<MemberAccessExpressionSyntax>(tree, "C.Test");
+            Assert.Equal("C C.op_Implicit(System.Func<System.Int32> intDelegate)", model.GetSymbolInfo(memberAccess).Symbol.ToTestDisplayString());  // Unexpected: Should be "void C.Test()"
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/ConversionTests.cs:485]: >// Unexpected: Should be "void C.Test()" [](http://example.com/codeflow?start=149&length=40)

Are we planning to address this?

File: src/Compilers/CSharp/Test/Symbol/Symbols/ConversionTests.cs (hunk 3 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Symbol/Symbols/ConversionTests.cs:442]: Done. There are some existing unexpected behaviors there. Filing issues for those

File: src/Compilers/VisualBasic/Test/Semantic/Semantics/Conversions.vb [hunk 7]
```diff
@@ -5154,5 +5154,33 @@ True
...
+End Class
+    ]]></file>
+    </compilation>)
+
+            compilation.AssertTheseEmitDiagnostics(<expected>
+BC30581: 'AddressOf' expression cannot be converted to 'C' because 'C' is not a delegate type.
+        Dim x As C = DirectCast(AddressOf C.Test, C)
+                                ~~~~~~~~~~~~~~~~
+</expected>)
+        End Sub
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/Semantics/Conversions.vb:5183]: >End Sub [](http://example.com/codeflow?start=8&length=7)

It doesn't look like we are testing SemanticModel behavior

**Comment by AlekseyTs** [comment]
Done with review pass (commit 2)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 4)

**Review by 333fred — APPROVED** [review]

**Pull request merged**